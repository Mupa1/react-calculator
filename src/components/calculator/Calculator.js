import React, { useState } from 'react';

import Display from '../display/Display';
import ButtonPanel from '../buttonPanel/ButtonPanel';
import calculate from '../../logic/calculate';
import styles from './Calculator.module.css';

const Calculator = () => {
  const [state, setState] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const handleClick = buttonName => {
    const result = calculate(state, buttonName);
    setState(result);
  };

  const { total, next } = state;
  const result = next || total || '0';

  return (
    <div className={styles.container}>
      <Display result={result} />
      <ButtonPanel clickHandler={handleClick} />
    </div>
  );
};

export default Calculator;
