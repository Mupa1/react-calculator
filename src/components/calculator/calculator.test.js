import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react';

import Calculator from './Calculator';

it('renders correctly', () => {
  const tree = renderer
    .create(<Calculator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

const clickButton = buttons => {
  const value = '';
  buttons.forEach(button => {
    Object.assign(value, fireEvent.click(screen.getByText(button)));
  });
  return value;
};

const total = resultTotal => {
  expect(screen.getByRole('heading')).toHaveTextContent(resultTotal);
};

describe('Calculator', () => {
  test('addition', () => {
    render(<Calculator />);
    clickButton(['6', '+', '7', '=']);
    total('13');
  });

  test('multiplication', () => {
    render(<Calculator />);
    clickButton(['9', 'x', '8', '=']);
    total('72');
  });

  test('division', () => {
    render(<Calculator />);
    clickButton(['8', 'รท', '4', '=']);
    total('2');
  });

  test('subtraction', () => {
    render(<Calculator />);
    clickButton(['8', '-', '4', '=']);
    total('4');
  });

  test('calculate numbers with decimals', () => {
    render(<Calculator />);
    clickButton(['1', '.', '2', '+', '1', '.', '2', '=']);
    total('2.4');
  });

  test('calculate modulus', () => {
    render(<Calculator />);
    clickButton(['1', '0', '0', '%']);
    total('1');
  });

  test('+/- should toggle from positive to negative number', () => {
    render(<Calculator />);
    clickButton(['7', '+/-']);
    total('-7');
  });

  test('+/- should toggle from negative and positive number', () => {
    render(<Calculator />);
    clickButton(['7', '+/-', '+/-']);
    total('7');
  });

  test('clear the operation when AC is pressed', () => {
    render(<Calculator />);
    clickButton(['7', '+/-', '+/-', 'AC']);
    total('0');
  });
});
