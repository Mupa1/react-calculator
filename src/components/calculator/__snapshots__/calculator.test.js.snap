// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="display"
  >
    <h3
      className="result"
    >
      0
    </h3>
  </div>
  <div
    className="row"
  >
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="button false orange"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
  </div>
  <div
    className="row"
  >
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="button false orange"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
  </div>
  <div
    className="row"
  >
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="button false orange"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
  </div>
  <div
    className="row"
  >
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="button false orange"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
  </div>
  <div
    className="row"
  >
    <button
      className="button wide gray"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      className="button false gray"
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="button false orange"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</div>
`;
